source/base/function_spherical.cc

```diff
@@ -0,0 +1,303 @@
+            sin(theta)=sin_theta
+            ],
+      H3);
+
+  where (used later to do tensor products):
+
+  ur     : [cos(theta)*sin(phi), sin_theta*sin(phi), cos_phi];
+  utheta : [-sin_theta, cos(theta), 0];
+  uphi   : [cos(theta)*cos_phi, sin_theta*cos_phi, -sin(phi)];
+
+   */
+
+  namespace
+  {
+    /**
+     * calculates out[i][j] = in1[i]in2[j]+in1[j]in2[i]
+     */
+    template <int dim>
+    void symm_outer_product(SymmetricTensor<2,dim> &out,
+                            const Tensor<1,dim> &in1,
+                            const Tensor<1,dim> &in2)
+    {
+      for (unsigned int i = 0; i < dim; i++)
+        for (unsigned int j = i; j < dim; j++)
+          out[i][j] = in1[i]*in2[j]+in1[j]*in2[i];
+    }
+
+    /**
+     * calculates out[i][j] = in[i]in[j]
+     */
+    template <int dim>
+    void symm_outer_product(SymmetricTensor<2,dim> &out,
+                            const Tensor<1,dim> &in)
+    {
+      for (unsigned int i = 0; i < dim; i++)
+        for (unsigned int j = i; j < dim; j++)
+          out[i][j] = in[i]*in[j];
+    }
+  }
+
+  template <int dim>
+  Spherical<dim>::Spherical(const Point<dim> &p)
+    :
+    Function<dim>(1),
+    coordinate_system_offset(p)
+  {
+  }
+
+  template <int dim>
+  Spherical<dim>::~Spherical()
+  {
+  }
+
+
+  template <int dim>
+  double
+  Spherical<dim>::value (const Point<dim>   &p_,
+                         const unsigned int) const
+  {
+    AssertThrow(dim==3,
+                ExcNotImplemented());
+
+    const Point<dim> p = p_ - coordinate_system_offset;
+    const std_cxx11::array<double, dim> sp = GeometricUtilities::Coordinates::to_spherical(p);
+    return svalue(sp);
+  }
+
+
+  template <int dim>
+  void
+  Spherical<dim>::set_unit_vectors(Tensor<1,dim> &unit_r,
+                                   Tensor<1,dim> &unit_theta,
+                                   Tensor<1,dim> &unit_phi,
+                                   const double &cos_theta,
+                                   const double &sin_theta,
+                                   const double &cos_phi,
+                                   const double &sin_phi) const
+  {
+    unit_r[0]     = cos_theta * sin_phi;
+    unit_r[1]     = sin_theta * sin_phi;
+    unit_r[2]     = cos_phi;
+
+    unit_theta[0] = -sin_theta;
+    unit_theta[1] =  cos_theta;
+    unit_theta[2] = 0.;
+
+    unit_phi[0]   = cos_theta * cos_phi;
+    unit_phi[1]   = sin_theta * cos_phi;
+    unit_phi[2]   = -sin_phi;
+  }
+
+
+  template <int dim>
+  Tensor<1,dim>
+  Spherical<dim>::gradient (const Point<dim>   &p_,
+                            const unsigned int) const
+  {
+    AssertThrow(dim==3,
+                ExcNotImplemented());
+
+    const Point<dim> p = p_ - coordinate_system_offset;
+    const std_cxx11::array<double, dim> sp = GeometricUtilities::Coordinates::to_spherical(p);
+    const Tensor<1,dim> sg = sgradient(sp);
+
+    // FIXME: somewhat backwards, but we need cos/sin's for unit vectors
```

Comment:
I suppose that if you're not going to address this now, then its probably best to remove this comment. Same goes for the comment in the Hessian calculation.